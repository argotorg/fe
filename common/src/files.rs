use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use std::{
    collections::HashMap,
    convert::TryInto,
    fs,
    io,
    ops::Range,
    path::Path,
};

pub struct SourceFile {
    id: SourceFileId,
    name: String,
    content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash)]
pub struct SourceFileId(u128);

impl SourceFile {
    pub fn new(name: String, content: String) -> Self {
        let hash = keccak::full_as_bytes(content.as_bytes());
        let line_starts = cs::files::line_starts(&content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name,
            content,
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        Some(Span::new(
            *self.line_starts.get(line_index)?,
            *self.line_starts.get(line_index + 1)?,
        ))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    files: HashMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: HashMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: HashMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: String, content: String) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    pub fn load_file(&mut self, path: String) -> io::Result<(SourceFileId, String)> {
        let content = self.loader.load_file(&Path::new(&path))?;
        let id = self.add_file(path, content.clone());
        Ok((id, content))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}
