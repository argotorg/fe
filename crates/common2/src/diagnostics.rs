use parser::ast::SyntaxNodePtr;

use crate::InputFile;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CompleteDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
    pub sub_diagnostics: Vec<SubDiagnostic>,
    pub error_code: GlobalErrorCode,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct GlobalErrorCode {
    pub pass: AnalysisPass,
    pub local_code: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SubDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Span {
    pub file: InputFile,
    pub node: SyntaxNodePtr,
    pub origin: SpanOrigin,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum Severity {
    Error,
    Warning,
    Note,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum SpanOrigin {
    Raw,
    Expanded,
    Desugared,
}

#[repr(u16)]
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum AnalysisPass {
    Parse = 1,
    NameResolution,
    TyCheck,

    ExternalAnalysis(ExternalAnalysisKey) = u16::MAX,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ExternalAnalysisKey {
    name: String,
}
