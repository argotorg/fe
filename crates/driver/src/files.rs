use camino::Utf8PathBuf;

pub const FE_TOML: &str = "fe.toml";

pub fn find_project_root() -> Option<Utf8PathBuf> {
    let mut path = Utf8PathBuf::from_path_buf(
        std::env::current_dir().expect("Unable to get current directory"),
    )
    .expect("Expected utf8 path");

    loop {
        let fe_toml = path.join(FE_TOML);
        if fe_toml.is_file() {
            return Some(path);
        }

        if !path.pop() {
            break;
        }
    }

    None
}
