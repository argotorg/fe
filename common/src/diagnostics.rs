use crate::files::{FileStore, SourceFileId};
pub use codespan_reporting::diagnostic::{Diagnostic as CsDiagnostic, Label, LabelStyle, Severity};
use codespan_reporting::term;
use term::termcolor::{BufferWriter, ColorChoice};

pub type Diagnostic = CsDiagnostic<SourceFileId>;

/// Print the given diagnostics to stderr.
pub fn print_diagnostics(diagnostics: &[Diagnostic], files: &FileStore) {
    let writer = BufferWriter::stderr(ColorChoice::Auto);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).unwrap();
    }
    // If we use `writer` here, the output won't be captured by rust's test system.
    eprintln!("{}", std::str::from_utf8(buffer.as_slice()).unwrap());
}

/// Format the given diagnostics as a string.
pub fn diagnostics_string(diagnostics: &[Diagnostic], files: &FileStore) -> String {
    let writer = BufferWriter::stderr(ColorChoice::Never);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).expect("failed to emit diagnostic");
    }
    std::str::from_utf8(buffer.as_slice()).unwrap().to_string()
}
