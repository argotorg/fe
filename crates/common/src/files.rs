use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use include_dir::Dir;
use indexmap::indexmap;
use indexmap::IndexMap;
use smol_str::SmolStr;
use std::collections::BTreeMap;
use std::ops::Range;
use std::path::Path;
use std::{fs, io};

#[derive(PartialEq, Clone, Eq, Hash, Debug)]
pub struct SourceFile {
    pub id: SourceFileId,
    pub name: String,
    pub content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash, Debug, PartialOrd, Ord, Default)]
pub struct SourceFileId(pub u128);

impl SourceFile {
    pub fn new(name: &str, content: &str) -> Self {
        // Canonicalize new line character for Windows.
        let content_canonicalized = content.replace("\r\n", "\n");
        let hash = keccak::full_as_bytes(content_canonicalized.as_bytes());

        let line_starts = cs::files::line_starts(content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name: name.to_string(),
            content: content.to_string(),
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        let end = if line_index == self.line_starts.len() - 1 {
            self.content.len()
        } else {
            *self.line_starts.get(line_index + 1)?
        };
        Some(Span::new(self.id, *self.line_starts.get(line_index)?, end))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    pub files: BTreeMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: BTreeMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: BTreeMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: &str, content: &str) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    /// Adds an included dir to the file store.
    pub fn add_included_dir(&mut self, dir: &Dir) -> Vec<SourceFileId> {
        let mut file_ids = vec![];

        for file in dir.files() {
            file_ids.push(
                self.add_file(
                    file.path()
                        .to_str()
                        .expect("cannot convert file path to string"),
                    file.contents_utf8()
                        .expect("could not get utf8 encoded file content"),
                ),
            );
        }

        for sub_dir in dir.dirs() {
            file_ids.extend(self.add_included_dir(sub_dir))
        }

        file_ids
    }

    /// Adds the included libraries to the file store and returns a mapping of
    /// library names to file ids.
    pub fn add_included_libraries(&mut self) -> IndexMap<SmolStr, Vec<SourceFileId>> {
        indexmap! {
            "std".into() => self.add_included_dir(&fe_library::STD)
        }
    }

    pub fn load_file(&mut self, path: &str) -> io::Result<(String, SourceFileId)> {
        let content = self.loader.load_file(Path::new(&path))?;
        let id = self.add_file(path, &content);
        Ok((content, id))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }

    pub fn all_files(&self) -> Vec<SourceFileId> {
        self.files.keys().copied().collect()
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}

impl Default for FileStore {
    fn default() -> Self {
        Self::new()
    }
}
