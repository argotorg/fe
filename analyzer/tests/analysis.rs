use fe_analyzer;
use fe_analyzer::context::Context;
use fe_common::diagnostics::{diagnostics_string, Diagnostic, Label};
use fe_common::files::{FileStore, SourceFileId};
use fe_parser::node::Span;
use insta::assert_snapshot;
use rstest::rstest;
use std::collections::hash_map::DefaultHasher;
use std::fmt::Debug;
use std::fs;
use std::hash::{Hash, Hasher};

#[rstest(
    fixture,
    case::erc20_token("demos/erc20_token.fe"),
    case::guest_book("demos/guest_book.fe"),
    case::uniswap("demos/uniswap.fe"),
    case::address_bytes10_map("features/address_bytes10_map.fe"),
    case::assert("features/assert.fe"),
    case::aug_assign("features/aug_assign.fe"),
    case::base_tuple("features/base_tuple.fe"),
    case::call_statement_with_args("features/call_statement_with_args.fe"),
    case::call_statement_with_args_2("features/call_statement_with_args_2.fe"),
    case::call_statement_without_args("features/call_statement_without_args.fe"),
    case::checked_arithmetic("features/checked_arithmetic.fe"),
    case::constructor("features/constructor.fe"),
    case::create2_contract("features/create2_contract.fe"),
    case::create_contract("features/create_contract.fe"),
    case::create_contract_from_init("features/create_contract_from_init.fe"),
    case::empty("features/empty.fe"),
    case::events("features/events.fe"),
    case::external_contract("features/external_contract.fe"),
    case::for_loop_with_break("features/for_loop_with_break.fe"),
    case::for_loop_with_continue("features/for_loop_with_continue.fe"),
    case::for_loop_with_static_array("features/for_loop_with_static_array.fe"),
    case::if_statement("features/if_statement.fe"),
    case::if_statement_2("features/if_statement_2.fe"),
    case::if_statement_with_block_declaration("features/if_statement_with_block_declaration.fe"),
    case::keccak("features/keccak.fe"),
    case::math("features/math.fe"),
    case::multi_param("features/multi_param.fe"),
    case::nested_map("features/nested_map.fe"),
    case::numeric_sizes("features/numeric_sizes.fe"),
    case::ownable("features/ownable.fe"),
    case::return_addition_i256("features/return_addition_i256.fe"),
    case::return_addition_u128("features/return_addition_u128.fe"),
    case::return_addition_u256("features/return_addition_u256.fe"),
    case::return_array("features/return_array.fe"),
    case::return_bitwiseand_u128("features/return_bitwiseand_u128.fe"),
    case::return_bitwiseand_u256("features/return_bitwiseand_u256.fe"),
    case::return_bitwiseor_u256("features/return_bitwiseor_u256.fe"),
    case::return_bitwiseshl_u256("features/return_bitwiseshl_u256.fe"),
    case::return_bitwiseshr_i256("features/return_bitwiseshr_i256.fe"),
    case::return_bitwiseshr_u256("features/return_bitwiseshr_u256.fe"),
    case::return_bitwisexor_u256("features/return_bitwisexor_u256.fe"),
    case::return_bool_false("features/return_bool_false.fe"),
    case::return_bool_inverted("features/return_bool_inverted.fe"),
    case::return_bool_op_and("features/return_bool_op_and.fe"),
    case::return_bool_op_or("features/return_bool_op_or.fe"),
    case::return_bool_true("features/return_bool_true.fe"),
    case::return_builtin_attributes("features/return_builtin_attributes.fe"),
    case::return_division_i256("features/return_division_i256.fe"),
    case::return_division_u256("features/return_division_u256.fe"),
    case::return_empty_tuple("features/return_empty_tuple.fe"),
    case::return_eq_u256("features/return_eq_u256.fe"),
    case::return_gt_i256("features/return_gt_i256.fe"),
    case::return_gt_u256("features/return_gt_u256.fe"),
    case::return_gte_i256("features/return_gte_i256.fe"),
    case::return_gte_u256("features/return_gte_u256.fe"),
    case::return_i128_cast("features/return_i128_cast.fe"),
    case::return_i256("features/return_i256.fe"),
    case::return_identity_u8("features/return_identity_u8.fe"),
    case::return_identity_u16("features/return_identity_u16.fe"),
    case::return_identity_u32("features/return_identity_u32.fe"),
    case::return_identity_u64("features/return_identity_u64.fe"),
    case::return_identity_u128("features/return_identity_u128.fe"),
    case::return_identity_u256("features/return_identity_u256.fe"),
    case::return_lt_i256("features/return_lt_i256.fe"),
    case::return_lt_u128("features/return_lt_u128.fe"),
    case::return_lt_u256("features/return_lt_u256.fe"),
    case::return_lte_i256("features/return_lte_i256.fe"),
    case::return_lte_u256("features/return_lte_u256.fe"),
    case::return_mod_i256("features/return_mod_i256.fe"),
    case::return_mod_u256("features/return_mod_u256.fe"),
    case::return_msg_sig("features/return_msg_sig.fe"),
    case::return_multiplication_i256("features/return_multiplication_i256.fe"),
    case::return_multiplication_u256("features/return_multiplication_u256.fe"),
    case::return_noteq_u256("features/return_noteq_u256.fe"),
    case::return_pow_i256("features/return_pow_i256.fe"),
    case::return_pow_u256("features/return_pow_u256.fe"),
    case::return_subtraction_i256("features/return_subtraction_i256.fe"),
    case::return_subtraction_u256("features/return_subtraction_u256.fe"),
    case::return_u128_cast("features/return_u128_cast.fe"),
    case::return_u256("features/return_u256.fe"),
    case::return_u256_from_called_fn("features/return_u256_from_called_fn.fe"),
    case::return_u256_from_called_fn_with_args("features/return_u256_from_called_fn_with_args.fe"),
    case::revert("features/revert.fe"),
    case::self_address("features/self_address.fe"),
    case::sized_vals_in_sto("features/sized_vals_in_sto.fe"),
    case::strings("features/strings.fe"),
    case::structs("features/structs.fe"),
    case::ternary_expression("features/ternary_expression.fe"),
    case::two_contracts("features/two_contracts.fe"),
    case::u8_u8_map("features/u8_u8_map.fe"),
    case::u16_u16_map("features/u16_u16_map.fe"),
    case::u32_u32_map("features/u32_u32_map.fe"),
    case::u64_u64_map("features/u64_u64_map.fe"),
    case::u128_u128_map("features/u128_u128_map.fe"),
    case::u256_u256_map("features/u256_u256_map.fe"),
    case::while_loop("features/while_loop.fe"),
    case::while_loop_with_break("features/while_loop_with_break.fe"),
    case::while_loop_with_break_2("features/while_loop_with_break_2.fe"),
    case::while_loop_with_continue("features/while_loop_with_continue.fe"),
    case::abi_encoding_stress("stress/abi_encoding_stress.fe"),
    case::data_copying_stress("stress/data_copying_stress.fe"),
    case::tuple_stress("stress/tuple_stress.fe")
)]
fn analysis(fixture: &str) {
    let src = &fs::read_to_string(format!("tests/fixtures/{}", fixture)).expect("");
    let fe_module = fe_parser::parse_file(src, SourceFileId(0)).expect("").0;

    let context = fe_analyzer::analyze(&fe_module).expect("");

    assert_snapshot!(build_snapshot(fixture, src, &context))
}

fn build_snapshot(path: &str, src: &str, context: &Context) -> String {
    let mut file_store = FileStore::new();
    let id = file_store.add_file(path, src);

    let diagnostics = [
        build_diagnostics(id, &context.get_spanned_expressions()),
        build_diagnostics(id, &context.get_spanned_emits()),
        build_diagnostics(id, &context.get_spanned_functions()),
        build_diagnostics(id, &context.get_spanned_declarations()),
        build_diagnostics(id, &context.get_spanned_contracts()),
        build_diagnostics(id, &context.get_spanned_calls()),
        build_diagnostics(id, &context.get_spanned_events()),
        build_diagnostics(id, &context.get_spanned_type_descs()),
    ]
    .concat();

    format!(
        "{:#?}\n\n{}",
        context
            .get_module()
            .expect("context is missing module attributes"),
        diagnostics_string(&diagnostics, &file_store)
    )
}

fn build_diagnostics<T: Hash + Debug>(
    file_id: SourceFileId,
    spanned_attributes: &[(Span, T)],
) -> Vec<Diagnostic> {
    let diagnostics = spanned_attributes
        .into_iter()
        .map(|(span, attributes)| build_attributes_diagnostic(file_id, span, attributes))
        .collect::<Vec<_>>();

    diagnostics
}

fn build_attributes_diagnostic<T: Hash + Debug>(
    file_id: SourceFileId,
    span: &Span,
    attributes: &T,
) -> Diagnostic {
    // Hash the attributes and label the span with it.
    let label = Label::primary(file_id, span.start..span.end)
        .with_message(format!("attributes hash: {}", hash(attributes)));
    Diagnostic::note()
        .with_labels(vec![label])
        .with_notes(vec![format!("{:#?}", attributes)])
}

fn hash<T: Hash>(item: &T) -> u64 {
    let mut s = DefaultHasher::new();
    item.hash(&mut s);
    s.finish()
}
