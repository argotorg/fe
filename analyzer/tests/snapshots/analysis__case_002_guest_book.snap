---
source: analyzer/tests/analysis.rs
expression: "build_snapshot(fixture, src, &context)"

---
ModuleAttributes {
    type_defs: {
        "BookMsg": Array(
            Array {
                size: 100,
                inner: Byte,
            },
        ),
        "GuestBook": Contract(
            Contract {
                name: "GuestBook",
                functions: [
                    FunctionAttributes {
                        is_public: true,
                        name: "get_msg",
                        params: [
                            (
                                "addr",
                                Base(
                                    Address,
                                ),
                            ),
                        ],
                        return_type: Array(
                            Array {
                                size: 100,
                                inner: Byte,
                            },
                        ),
                    },
                    FunctionAttributes {
                        is_public: true,
                        name: "sign",
                        params: [
                            (
                                "book_msg",
                                Array(
                                    Array {
                                        size: 100,
                                        inner: Byte,
                                    },
                                ),
                            ),
                        ],
                        return_type: Tuple(
                            Tuple {
                                items: [],
                            },
                        ),
                    },
                ],
            },
        ),
    },
    tuples_used: {},
}

note: 
   ┌─ demos/guest_book.fe:10:9
   │
10 │         self.guest_book[msg.sender] = book_msg
   │         ^^^^^^^^^^^^^^^ attributes hash: 7171665311786934563
   │
   = ExpressionAttributes {
         typ: Map(
             Map {
                 key: Address,
                 value: Array(
                     Array {
                         size: 100,
                         inner: Byte,
                     },
                 ),
             },
         ),
         location: Storage {
             nonce: Some(
                 0,
             ),
         },
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:10:25
   │
10 │         self.guest_book[msg.sender] = book_msg
   │                         ^^^^^^^^^^ attributes hash: 513065555763557710
   │
   = ExpressionAttributes {
         typ: Base(
             Address,
         ),
         location: Value,
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:10:9
   │
10 │         self.guest_book[msg.sender] = book_msg
   │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ attributes hash: 453830748720542459
   │
   = ExpressionAttributes {
         typ: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
         location: Storage {
             nonce: None,
         },
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:10:39
   │
10 │         self.guest_book[msg.sender] = book_msg
   │                                       ^^^^^^^^ attributes hash: 6578844474441717363
   │
   = ExpressionAttributes {
         typ: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
         location: Memory,
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:12:30
   │
12 │         emit Signed(book_msg=book_msg)
   │                              ^^^^^^^^ attributes hash: 6578844474441717363
   │
   = ExpressionAttributes {
         typ: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
         location: Memory,
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:15:16
   │
15 │         return self.guest_book[addr].to_mem()
   │                ^^^^^^^^^^^^^^^ attributes hash: 7171665311786934563
   │
   = ExpressionAttributes {
         typ: Map(
             Map {
                 key: Address,
                 value: Array(
                     Array {
                         size: 100,
                         inner: Byte,
                     },
                 ),
             },
         ),
         location: Storage {
             nonce: Some(
                 0,
             ),
         },
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:15:32
   │
15 │         return self.guest_book[addr].to_mem()
   │                                ^^^^ attributes hash: 513065555763557710
   │
   = ExpressionAttributes {
         typ: Base(
             Address,
         ),
         location: Value,
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:15:16
   │
15 │         return self.guest_book[addr].to_mem()
   │                ^^^^^^^^^^^^^^^^^^^^^ attributes hash: 453830748720542459
   │
   = ExpressionAttributes {
         typ: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
         location: Storage {
             nonce: None,
         },
         move_location: None,
     }

note: 
   ┌─ demos/guest_book.fe:15:16
   │
15 │         return self.guest_book[addr].to_mem()
   │                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attributes hash: 16413446491870513894
   │
   = ExpressionAttributes {
         typ: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
         location: Storage {
             nonce: None,
         },
         move_location: Some(
             Memory,
         ),
     }

note: 
   ┌─ demos/guest_book.fe:12:9
   │
12 │         emit Signed(book_msg=book_msg)
   │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attributes hash: 10823436434315354262
   │
   = EventDef {
         name: "Signed",
         topic: "0xf95318ba442251854c1277ed370e2adb2cd6dc2156bdffdd75dc5d798b31ab0f",
         fields: [
             (
                 "book_msg",
                 Array(
                     Array {
                         size: 100,
                         inner: Byte,
                     },
                 ),
             ),
         ],
         indexed_fields: [],
     }

note: 
   ┌─ demos/guest_book.fe:9:5
   │  
 9 │ ╭     pub def sign(book_msg: BookMsg):
10 │ │         self.guest_book[msg.sender] = book_msg
11 │ │ 
12 │ │         emit Signed(book_msg=book_msg)
   │ ╰──────────────────────────────────────^ attributes hash: 5066331849586951910
   │  
   = FunctionAttributes {
         is_public: true,
         name: "sign",
         params: [
             (
                 "book_msg",
                 Array(
                     Array {
                         size: 100,
                         inner: Byte,
                     },
                 ),
             ),
         ],
         return_type: Tuple(
             Tuple {
                 items: [],
             },
         ),
     }

note: 
   ┌─ demos/guest_book.fe:14:5
   │  
14 │ ╭     pub def get_msg(addr: address) -> BookMsg:
15 │ │         return self.guest_book[addr].to_mem()
   │ ╰─────────────────────────────────────────────^ attributes hash: 14720587159970812425
   │  
   = FunctionAttributes {
         is_public: true,
         name: "get_msg",
         params: [
             (
                 "addr",
                 Base(
                     Address,
                 ),
             ),
         ],
         return_type: Array(
             Array {
                 size: 100,
                 inner: Byte,
             },
         ),
     }

note: 
   ┌─ demos/guest_book.fe:3:1
   │  
 3 │ ╭ contract GuestBook:
 4 │ │     pub guest_book: map<address, BookMsg>
 5 │ │ 
 6 │ │     event Signed:
   · │
14 │ │     pub def get_msg(addr: address) -> BookMsg:
15 │ │         return self.guest_book[addr].to_mem()
   │ ╰─────────────────────────────────────────────^ attributes hash: 6562880586223247611
   │  
   = ContractAttributes {
         public_functions: [
             FunctionAttributes {
                 is_public: true,
                 name: "get_msg",
                 params: [
                     (
                         "addr",
                         Base(
                             Address,
                         ),
                     ),
                 ],
                 return_type: Array(
                     Array {
                         size: 100,
                         inner: Byte,
                     },
                 ),
             },
             FunctionAttributes {
                 is_public: true,
                 name: "sign",
                 params: [
                     (
                         "book_msg",
                         Array(
                             Array {
                                 size: 100,
                                 inner: Byte,
                             },
                         ),
                     ),
                 ],
                 return_type: Tuple(
                     Tuple {
                         items: [],
                     },
                 ),
             },
         ],
         init_function: None,
         events: [
             EventDef {
                 name: "Signed",
                 topic: "0xf95318ba442251854c1277ed370e2adb2cd6dc2156bdffdd75dc5d798b31ab0f",
                 fields: [
                     (
                         "book_msg",
                         Array(
                             Array {
                                 size: 100,
                                 inner: Byte,
                             },
                         ),
                     ),
                 ],
                 indexed_fields: [],
             },
         ],
         list_expressions: {},
         string_literals: {},
         structs: [],
         external_contracts: [],
         created_contracts: {},
     }

note: 
   ┌─ demos/guest_book.fe:15:16
   │
15 │         return self.guest_book[addr].to_mem()
   │                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attributes hash: 18270091135093349626
   │
   = ValueAttribute

note: 
  ┌─ demos/guest_book.fe:6:5
  │  
6 │ ╭     event Signed:
7 │ │         book_msg: BookMsg
  │ ╰─────────────────────────^ attributes hash: 10823436434315354262
  │  
  = EventDef {
        name: "Signed",
        topic: "0xf95318ba442251854c1277ed370e2adb2cd6dc2156bdffdd75dc5d798b31ab0f",
        fields: [
            (
                "book_msg",
                Array(
                    Array {
                        size: 100,
                        inner: Byte,
                    },
                ),
            ),
        ],
        indexed_fields: [],
    }

note: 
  ┌─ demos/guest_book.fe:7:19
  │
7 │         book_msg: BookMsg
  │                   ^^^^^^^ attributes hash: 5134157185158150929
  │
  = Array(
        Array {
            size: 100,
            inner: Byte,
        },
    )

note: 
  ┌─ demos/guest_book.fe:9:28
  │
9 │     pub def sign(book_msg: BookMsg):
  │                            ^^^^^^^ attributes hash: 5134157185158150929
  │
  = Array(
        Array {
            size: 100,
            inner: Byte,
        },
    )

note: 
   ┌─ demos/guest_book.fe:14:27
   │
14 │     pub def get_msg(addr: address) -> BookMsg:
   │                           ^^^^^^^ attributes hash: 574436082528610497
   │
   = Base(
         Address,
     )

note: 
   ┌─ demos/guest_book.fe:14:39
   │
14 │     pub def get_msg(addr: address) -> BookMsg:
   │                                       ^^^^^^^ attributes hash: 5134157185158150929
   │
   = Array(
         Array {
             size: 100,
             inner: Byte,
         },
     )

note: 
  ┌─ demos/guest_book.fe:4:21
  │
4 │     pub guest_book: map<address, BookMsg>
  │                     ^^^^^^^^^^^^^^^^^^^^^ attributes hash: 10049485715821994445
  │
  = Map(
        Map {
            key: Address,
            value: Array(
                Array {
                    size: 100,
                    inner: Byte,
                },
            ),
        },
    )


